src/mmc.c

```diff
@@ -0,0 +1,280 @@
+    *value = (int)strtol(buffer, NULL, 0);
+    DEBUG(PLUGIN_NAME "(%s): [%s]=%s (%d)", dev_name, MMC_MANUFACTOR, buffer,
+          *value);
+    return 0;
+  }
+
+  WARNING(PLUGIN_NAME "(%s): Unable to read manufacturer identifier (manfid)",
+          dev_name);
+  return EXIT_FAILURE;
+}
+
+static int mmc_read_oemid(const char *dev_name, int *value) {
+  char buffer[4096];
+
+  if (mmc_read_dev_attr(dev_name, MMC_OEM_ID, buffer, sizeof(buffer)) == 0) {
+    *value = (int)strtol(buffer, NULL, 0);
+    DEBUG(PLUGIN_NAME "(%s): [%s]=%s (%d)", dev_name, MMC_OEM_ID, buffer,
+          *value);
+    return 0;
+  }
+
+  WARNING(
+      PLUGIN_NAME
+      "(%s): Unable to read original equipment manufacturer identifier (oemid)",
+      dev_name);
+  return EXIT_FAILURE;
+}
+
+enum mmc_manfid {
+  MANUFACTUR_SWISSBIT = 93, // 0x5d
+};
+
+enum mmc_oemid_swissbit {
+  OEMID_SWISSBIT_1 = 21314, // 0x5342
+};
+
+#define MMC_POWER_CYCLES "mmc_power_cycles"
+#define MMC_BLOCK_ERASES "mmc_block_erases"
+#define MMC_BAD_BLOCKS "mmc_bad_blocks"
+
+// Size of string buffer with '\0'
+#define SWISSBIT_LENGTH_SPARE_BLOCKS 3
+#define SWISSBIT_LENGTH_BLOCK_ERASES 13
+#define SWISSBIT_LENGTH_POWER_ON 9
+
+#define SWISSBIT_SSR_START_SPARE_BLOCKS 66
+#define SWISSBIT_SSR_START_BLOCK_ERASES 92
+#define SWISSBIT_SSR_START_POWER_ON 112
+
+static int mmc_read_ssr_swissbit(const char *dev_name) {
+  char buffer[4096];
+  int oemid;
+  int value;
+  int length;
+  char bad_blocks[SWISSBIT_LENGTH_SPARE_BLOCKS];
+  char block_erases[SWISSBIT_LENGTH_BLOCK_ERASES];
+  char power_on[SWISSBIT_LENGTH_POWER_ON];
+
+  if (mmc_read_oemid(dev_name, &oemid) != 0) {
+    return EXIT_FAILURE;
+  }
+
+  if (oemid != OEMID_SWISSBIT_1) {
+    INFO(PLUGIN_NAME
+         "(%s): The mmc device is not suppored by this plugin (oemid: 0x%x)",
+         dev_name, oemid);
+    return EXIT_FAILURE;
+  }
+
+  if (mmc_read_dev_attr(dev_name, MMC_SSR, buffer, sizeof(buffer)) != 0) {
+    return EXIT_FAILURE;
+  }
+
+  /*
+   * Since the register is read out as a byte stream, it is 128 bytes long.
+   * One char represents a half byte (nibble).
+   *
+   */
+  length = strlen(buffer);
+  DEBUG(PLUGIN_NAME ": %d byte read from SSR register", length);
+  if (length < 128) {
+    INFO(PLUGIN_NAME "(%s): The SSR register is not 128 byte long", dev_name);
+    return EXIT_FAILURE;
+  }
+  DEBUG(PLUGIN_NAME "(%s): [%s]=%s", dev_name, MMC_SSR, buffer);
+
+  /* write MMC_BAD_BLOCKS */
+  strncpy(bad_blocks, &buffer[SWISSBIT_SSR_START_SPARE_BLOCKS],
+          sizeof(bad_blocks) - 1);
+  bad_blocks[sizeof(bad_blocks) - 1] = '\0';
+  value = (int)strtol(bad_blocks, NULL, 16);
+  /* convert to more common bad blocks information */
+  value = abs(value - 100);
+  DEBUG(PLUGIN_NAME "(%s): [bad_blocks] str=%s int=%d", dev_name, bad_blocks,
+        value);
+  mmc_submit(dev_name, MMC_BAD_BLOCKS, (gauge_t)value);
+
+  /* write MMC_BLOCK_ERASES */
+  strncpy(block_erases, &buffer[SWISSBIT_SSR_START_BLOCK_ERASES],
```

Comment:
can you please use sstrncpy here as well? Like above, taken from https://github.com/collectd/collectd/blob/main/src/utils/common/common.h#L67

Reply:
done, thanks